File: RoundRobin.py
# coding: utf-8
# Algoritmo de planificación por turnos (Round Robin):
# @faustinoaq Sept, 2016

from sys import argv


class Proceso():

    def __init__(self, pcb, id, instru, estado, posic):
        self.pcb = pcb
        self.id = id
        self.instru = instru
        self.estado = estado
        self.posic = posic


class RoundRobin():

    def __init__(self, n, quantum, archivo):
        self.n = n
        self.quantum = quantum
        self.archivo = archivo
        self.buffer = []
        self.procesos = [
            Proceso("P{}".format(i), "0{}".format(i), 100, "N", i)
            for i in range(1, self.n + 1)
        ]
        self.titulo_buffer()
        self.cola_listo()
        self.procesar_cola()
        self.guardar_archivo()

    def titulo_buffer(self):
        self.buffer = ["Algoritmo de planificación Round Robin\n"]
        self.buffer.append("N(Nuevo) L(Listo) E(Ejecución) T(Terminado)\n")
        self.buffer.append("Cantidad de procesos: {}\n".format(self.n))
        self.buffer.append("Quantum: {}\n".format(self.quantum))
        self.guardar_buffer()

    def guardar_buffer(self):
        self.buffer.append("\nProceso")
        for p in self.procesos:
            self.buffer.append("\t{}".format(p.pcb))
        self.buffer.append("\nID")
        for p in self.procesos:
            self.buffer.append("\t{}".format(p.id))
        self.buffer.append("\nInstruc")
        for p in self.procesos:
            self.buffer.append("\t{}".format(p.instru))
        self.buffer.append("\nEstado")
        for p in self.procesos:
            self.buffer.append("\t{}".format(p.estado))
        self.buffer.append("\nPosic")
        for p in self.procesos:
            self.buffer.append("\t{}".format(p.posic))
        self.buffer.append("\n")

    def cola_listo(self):
        for p in self.procesos:
            p.estado = "L"
        self.guardar_buffer()

    def procesar_cola(self):
        while self.procesos_no_terminados():
            for p in self.procesos:
                self.trabajar_proceso(p)
        self.reposicionar_cola()
        self.guardar_buffer()

    def reposicionar_cola(self):
        posicion = 0
        for p in self.procesos:
            if p.estado == "T":
                p.posic = 0
            else:
                posicion += 1
                p.posic = posicion

    def procesos_no_terminados(self):
        for p in self.procesos:
            if p.estado != "T":
                return True
        return False

    def trabajar_proceso(self, p):
        if p.instru > self.quantum:
            p.instru -= self.quantum
            p.estado = "E"
            self.guardar_buffer()
            p.estado = "T" if p.instru == 0 else "L"
        elif p.estado != "T":
            p.instru = 0
            p.estado = "E"
            self.reposicionar_cola()
            self.guardar_buffer()
            p.estado = "T"

    def guardar_archivo(self):
        try:
            with open(self.archivo, "w") as f:
                f.write("".join(self.buffer))
                f.close()
            print("{} guardado correctamente".format(self.archivo))
        except IOError:
            raise("Error al guardar {}".format(self.archivo))

RoundRobin(int(argv[1]), int(argv[2]), "pytest.txt")

File: RoundRobin.java
import java.io.FileWriter;
import java.io.BufferedWriter;
import java.io.PrintWriter;
import java.util.ArrayList;

/**
 * Algoritmo de planificación por turnos (Round Robin)
 * @faustinoaq Sept, 2016
 */

class Proceso {
    String pcb;
    String id;
    int instru;
    String estado;
    int posic;
}

class RoundRobin {
    int n;
    int quantum;
    String archivo;
    Proceso[] procesos;
    ArrayList<String> buffer;

    RoundRobin (int n, int quantum, String archivo) {
        this.n = n;
        this.quantum = quantum;
        this.archivo = archivo;
        this.procesos = new Proceso[n];
        this.buffer = new ArrayList<>();
        Proceso p;
        for (int i = 0; i < this.n; i++) {
            p = new Proceso();
            p.pcb = "P" + Integer.toString(i+1);
            p.id = "0" + Integer.toString(i+1);
            p.instru = 100;
            p.estado = "N";
            p.posic = i + 1;
            procesos[i] = p;
        }
        tituloBuffer();
        colaListo();
        procesarCola();
        System.out.println(this.archivo + " guardado correctamente");
    }

    public static void main(String[] args) {
        new RoundRobin(Integer.parseInt(args[0]),
                       Integer.parseInt(args[1]),
                       "jatest.txt");
    }

    private void tituloBuffer() {
        this.buffer.add("Algoritmo de planificación Round Robin\n");
        this.buffer.add("N(Nuevo) L(Listo) E(Ejecución) T(Terminado)\n");
        this.buffer.add("Cantidad de procesos: " + this.n + "\n");
        this.buffer.add("Quantum: " + this.quantum + "\n");
        guardarBuffer();
    }

    private void guardarBuffer() {
        this.buffer.add("\nProceso");
        for(Proceso p : procesos) {
            this.buffer.add("\t" + p.pcb);
        }
        this.buffer.add("\nID");
        for(Proceso p : procesos) {
            this.buffer.add("\t" + p.id);
        }
        this.buffer.add("\nInstruc");
        for(Proceso p : procesos) {
            this.buffer.add("\t" + Integer.toString(p.instru));
        }
        this.buffer.add("\nEstado");
        for(Proceso p : procesos) {
            this.buffer.add("\t" + p.estado);
        }
        this.buffer.add("\nPosic");
        for(Proceso p : procesos) {
            this.buffer.add("\t" + Integer.toString(p.posic));
        }
        this.buffer.add("\n");
        guardarArchivo();
    }

    private void colaListo() {
        for (int i = 0; i < this.n; i++) {
            this.procesos[i].estado = "L";
        }
        guardarBuffer();
    }

    private void procesarCola() {
        while (procesosNoTerminados()) {
            for (Proceso p : procesos) {
                trabajarProceso(p);
            } 
        }
        reposicionarCola();
        guardarBuffer();
    }

    private void reposicionarCola() {
        int posicion = 0;
        for (Proceso p : procesos) {
            if (p.estado.equals("T")) {
                p.posic = 0;
            } else {
                posicion += 1;
                p.posic = posicion;
            }
        }
    }

    private boolean procesosNoTerminados() {
        for (Proceso p : procesos) {
            if (!"T".equals(p.estado)) {
                return true;
            }
        }
        return false;
    }

    private void trabajarProceso(Proceso p) {
        if (p.instru > quantum) {
            p.instru -= quantum;
            p.estado = "E";
            guardarBuffer();
            p.estado = (p.instru == 0) ? "T" : "L";
        } else if (!"T".equals(p.estado)) {
            p.instru = 0;
            p.estado = "E";
            reposicionarCola();
            guardarBuffer();
            p.estado = "T";
        }
    }

    private void guardarArchivo() {
        try {
            String buf = String.join("", this.buffer);
            FileWriter fw = new FileWriter(this.archivo, true);
            BufferedWriter bw = new BufferedWriter(fw);
            PrintWriter pw = new PrintWriter(bw);
            pw.println(buf);
            pw.close();
            this.buffer.clear();
        } catch (Exception e) {
            System.out.println("Error al guardar " + this.archivo);
        }
    }
}
